File: grey-haven-plugins/grey-haven-hooks/hooks/js/subagent-work-validator.js
#!/usr/bin/env node

/**
 * Subagent Work Validator Hook for JavaScript/TypeScript
 * ========================================================
 * Type: SubagentStop, PostToolUse(Task)
 * Description: Validates that subagent completed its assigned task properly
 *
 * This hook ensures subagents have completed their work according to standards:
 * - Code compiles/runs
 * - Tests were added if code was written
 * - Documentation was updated
 * - No obvious issues introduced
 */

const fs = require('fs');
const path = require('path');
const { execSync } = require('child_process');

class SubagentWorkValidator {
    constructor(projectDir, transcriptPath = null) {
        this.projectDir = projectDir;
        this.transcriptPath = transcriptPath;
        this.issues = [];
        this.suggestions = [];
    }

    analyzeTranscript() {
        if (!this.transcriptPath || !fs.existsSync(this.transcriptPath)) {
            return {};
        }

        try {
            const content = fs.readFileSync(this.transcriptPath, 'utf8');
            
            // Look for indicators of incomplete work
            const indicators = {
                errors: (content.match(/error|Error|ERROR/g) || []).length,
                warnings: (content.match(/warning|Warning|WARNING/g) || []).length,
                todos: (content.match(/TODO|FIXME|XXX|HACK/g) || []).length,
                skipped: (content.match(/skip|Skip|SKIP|defer|Defer|DEFER/g) || []).length,
                failedTests: (content.match(/test.*fail|fail.*test|FAIL/gi) || []).length,
            };
            
            return indicators;
        } catch {
            return {};
        }
    }

    checkSyntaxErrors() {
        try {
            // Get recently modified files (within last 5 minutes)
            const recentFiles = execSync(
                `find "${this.projectDir}" -type f -mmin -5 2>/dev/null`,
                { encoding: 'utf8' }
            ).trim().split('\n').filter(Boolean);

            const filesWithErrors = [];

            for (const file of recentFiles) {
                // JavaScript/TypeScript syntax check
                if (file.endsWith('.js') || file.endsWith('.jsx')) {
                    try {
                        execSync(`node --check "${file}" 2>/dev/null`);
                    } catch {
                        filesWithErrors.push(path.basename(file));
                    }
                }
                
                // TypeScript check
                if (file.endsWith('.ts') || file.endsWith('.tsx')) {
                    // Check if TypeScript is available
                    try {
                        execSync('which tsc', { stdio: 'ignore' });
                        execSync(`tsc --noEmit "${file}" 2>/dev/null`);
                    } catch {
                        // Only report if tsc is available
                        if (fs.existsSync(path.join(this.projectDir, 'node_modules/typescript'))) {
                            filesWithErrors.push(path.basename(file));
                        }
                    }
                }
                
                // JSON syntax check
                if (file.endsWith('.json')) {
                    try {
                        JSON.parse(fs.readFileSync(file, 'utf8'));
                    } catch {
                        filesWithErrors.push(path.basename(file));
                    }
                }
            }

            if (filesWithErrors.length) {
                this.issues.push(`Syntax errors in: ${filesWithErrors.slice(0, 3).join(', ')}`);
                return false;
            }

            return true;
        } catch {
            return true; // Don't block if we can't check
        }
    }

    checkTestCoverage() {
        try {
            // Get files modified in last 5 minutes
            const modifiedFiles = execSync(
                `find "${this.projectDir}" -type f \\( -name "*.js" -o -name "*.jsx" -o -name "*.ts" -o -name "*.tsx" \\) -mmin -5 2>/dev/null`,
                { encoding: 'utf8' }
            ).trim().split('\n').filter(Boolean);

            if (!modifiedFiles.length) {
                return true;
            }

            const modifiedCodeFiles = [];
            const modifiedTestFiles = [];

            for (const file of modifiedFiles) {
                if (/\.test\.|\.spec\.|__tests__|test_/.test(file)) {
                    modifiedTestFiles.push(file);
                } else {
                    modifiedCodeFiles.push(file);
                }
            }

            // If code was modified but no tests, suggest adding tests
            if (modifiedCodeFiles.length && !modifiedTestFiles.length) {
                // Check if any of the modified files are significant (not just config)
                const significantFiles = modifiedCodeFiles.filter(f => 
                    !f.includes('config') && 
                    !f.includes('.d.ts') &&
                    !f.endsWith('package.json')
                );
                
                if (significantFiles.length) {
                    this.suggestions.push('Consider adding tests for the new code');
                }
            }

            return true; // Don't block, just suggest
        } catch {
            return true;
        }
    }

    checkDocumentation() {
        try {
            // Check if any significant code files were changed
            const codeFiles = execSync(
                `find "${this.projectDir}" -type f \\( -name "*.js" -o -name "*.jsx" -o -name "*.ts" -o -name "*.tsx" \\) -mmin -5 2>/dev/null`,
                { encoding: 'utf8' }
            ).trim().split('\n').filter(Boolean);

            const codeModified = codeFiles.length > 0;

            // Check if any docs were modified
            const docFiles = execSync(
                `find "${this.projectDir}" -type f \\( -name "*.md" -o -name "*.mdx" -o -path "*/docs/*" \\) -mmin -5 2>/dev/null`,
                { encoding: 'utf8' }
            ).trim().split('\n').filter(Boolean);

            const docsModified = docFiles.length > 0;

            if (codeModified && !docsModified) {
                // Check if the changes were significant
                try {
                    const diffStat = execSync('git diff --stat', {
                        encoding: 'utf8',
                        cwd: this.projectDir
                    });
                    
                    const insertions = diffStat.match(/(\d+) insertion/);
                    const totalInsertions = insertions ? parseInt(insertions[1]) : 0;
                    
                    if (totalInsertions > 20) {
                        this.suggestions.push('Documentation may need updating for these changes');
                    }
                } catch {}
            }

            // Check for missing JSDoc in new functions
            for (const file of codeFiles.slice(0, 3)) {
                try {
                    const content = fs.readFileSync(file, 'utf8');
                    
                    // Check for exported functions without JSDoc
                    const exportedFunctions = content.match(/export\s+(async\s+)?function\s+\w+/g) || [];
                    const jsdocComments = content.match(/\/\*\*[\s\S]*?\*\//g) || [];
                    
                    if (exportedFunctions.length > jsdocComments.length) {
                        this.suggestions.push(`Some exported functions lack JSDoc in ${path.basename(file)}`);
                        break;
                    }
                } catch {}
            }

            return true; // Don't block on documentation
        } catch {
            return true;
        }
    }

    checkImports() {
        try {
            // Check JavaScript/TypeScript files
            const jsFiles = execSync(
                `find "${this.projectDir}" -type f \\( -name "*.js" -o -name "*.jsx" -o -name "*.ts" -o -name "*.tsx" \\) -mmin -5 2>/dev/null`,
                { encoding: 'utf8' }
            ).trim().split('\n').filter(Boolean);

            for (const file of jsFiles) {
                if (!fs.existsSync(file)) continue;
                
                try {
                    const content = fs.readFileSync(file, 'utf8');
                    
                    // Check for missing imports (using undefined variables)
                    // This is a simple heuristic check
                    const imports = new Set();
                    const importMatches = content.matchAll(/import\s+(?:{[^}]+}|\w+|\*\s+as\s+\w+)\s+from/g);
                    for (const match of importMatches) {
                        const importText = match[0];
                        // Extract imported names
                        const names = importText.match(/\w+/g);
                        if (names) {
                            names.forEach(name => imports.add(name));
                        }
                    }
                    
                    // Check for duplicate imports
                    const importLines = content.match(/^import\s+.*$/gm) || [];
                    const importSources = importLines.map(line => {
                        const match = line.match(/from\s+['"](.*)['"]/);
                        return match ? match[1] : null;
                    }).filter(Boolean);
                    
                    const duplicates = importSources.filter((item, index) => 
                        importSources.indexOf(item) !== index
                    );
                    
                    if (duplicates.length) {
                        this.suggestions.push(`Duplicate imports in ${path.basename(file)}`);
                    }
                    
                    // Check for unused imports (basic check)
                    const unusedImports = [];
                    for (const imp of imports) {
                        if (imp === 'import' || imp === 'from' || imp === 'as') continue;
                        // Check if the import is used in the file (excluding the import line)
                        const regex = new RegExp(`\\b${imp}\\b`, 'g');
                        const matches = content.match(regex) || [];
                        if (matches.length <= 1) { // Only appears in import statement
                            unusedImports.push(imp);
                        }
                    }
                    
                    if (unusedImports.length > 3) {
                        this.suggestions.push(`Multiple unused imports in ${path.basename(file)}`);
                    }
                } catch {}
            }

            return true; // Don't block on import issues
        } catch {
            return true;
        }
    }

    checkPackageJson() {
        // Check if package.json was modified
        const packageJsonPath = path.join(this.projectDir, 'package.json');
        
        try {
            const stats = fs.statSync(packageJsonPath);
            const modifiedRecently = (Date.now() - stats.mtimeMs) < 5 * 60 * 1000; // 5 minutes
            
            if (modifiedRecently) {
                // Validate package.json
                try {
                    const content = fs.readFileSync(packageJsonPath, 'utf8');
                    JSON.parse(content); // Will throw if invalid
                    
                    // Check for lock file update
                    const hasNpmLock = fs.existsSync(path.join(this.projectDir, 'package-lock.json'));
                    const hasYarnLock = fs.existsSync(path.join(this.projectDir, 'yarn.lock'));
                    const hasPnpmLock = fs.existsSync(path.join(this.projectDir, 'pnpm-lock.yaml'));
                    const hasBunLock = fs.existsSync(path.join(this.projectDir, 'bun.lockb'));
                    
                    if (!hasNpmLock && !hasYarnLock && !hasPnpmLock && !hasBunLock) {
                        this.suggestions.push('package.json modified but no lock file found');
                    } else {
                        // Check if lock file is older than package.json
                        const lockFiles = [];
                        if (hasNpmLock) lockFiles.push('package-lock.json');
                        if (hasYarnLock) lockFiles.push('yarn.lock');
                        if (hasPnpmLock) lockFiles.push('pnpm-lock.yaml');
                        if (hasBunLock) lockFiles.push('bun.lockb');
                        
                        for (const lockFile of lockFiles) {
                            const lockPath = path.join(this.projectDir, lockFile);
                            const lockStats = fs.statSync(lockPath);
                            if (lockStats.mtimeMs < stats.mtimeMs) {
                                this.suggestions.push(`${lockFile} may need updating after package.json changes`);
                                break;
                            }
                        }
                    }
                } catch (error) {
                    this.issues.push('package.json is invalid JSON');
                }
            }
        } catch {}
        
        return true;
    }

    validate() {
        // Analyze transcript if available
        const transcriptIndicators = this.analyzeTranscript();
        
        // Check for serious issues
        this.checkSyntaxErrors();
        
        // Check for quality issues
        this.checkTestCoverage();
        this.checkDocumentation();
        this.checkImports();
        this.checkPackageJson();
        
        // Analyze transcript indicators
        if (transcriptIndicators.errors && transcriptIndicators.errors > 5) {
            this.issues.push('Multiple errors detected in subagent work');
        }
        if (transcriptIndicators.failedTests && transcriptIndicators.failedTests > 0) {
            this.issues.push('Tests appear to be failing');
        }
        if (transcriptIndicators.todos && transcriptIndicators.todos > 3) {
            this.suggestions.push('Several TODOs were left in the code');
        }
        
        // Determine if we should block
        if (this.issues.length > 0) {
            let reason = 'Subagent work validation failed:\n';
            reason += this.issues.map(issue => `❌ ${issue}`).join('\n');
            
            if (this.suggestions.length) {
                reason += '\n\nSuggestions:\n';
                reason += this.suggestions.map(suggestion => `💡 ${suggestion}`).join('\n');
            }
            
            return {
                decision: 'block',
                reason: reason + '\n\nPlease address these issues.'
            };
        } else if (this.suggestions.length) {
            // Just provide suggestions, don't block
            const feedback = 'Subagent work complete. Suggestions:\n' +
                           this.suggestions.map(suggestion => `💡 ${suggestion}`).join('\n');
            
            console.error(feedback);
            return {};
        } else {
            console.error('✅ Subagent work validated successfully!');
            return {};
        }
    }
}

// Main execution
let inputData = '';
process.stdin.on('data', chunk => inputData += chunk);
process.stdin.on('end', () => {
    try {
        const data = JSON.parse(inputData);
        
        // Handle both SubagentStop and PostToolUse(Task)
        const hookEvent = data.hook_event_name || '';
        
        if (hookEvent === 'PostToolUse') {
            const toolName = data.tool_name || '';
            if (toolName !== 'Task') {
                process.exit(0);
            }
        }
        
        const projectDir = data.cwd || process.cwd();
        const transcriptPath = data.transcript_path;
        
        const validator = new SubagentWorkValidator(projectDir, transcriptPath);
        const result = validator.validate();
        
        if (result.decision) {
            console.log(JSON.stringify(result));
        }
        
        process.exit(0);
    } catch (error) {
        // Don't block on errors
        process.exit(0);
    }
});

File: grey-haven-plugins/hooks/hooks/python/subagent-work-validator.py
#!/usr/bin/env -S uv run --quiet --script
# /// script
# dependencies = []
# ///
"""
Subagent Work Validator Hook
============================
Type: SubagentStop, PostToolUse(Task)
Description: Validates that subagent completed its assigned task properly

This hook ensures subagents have completed their work according to standards:
- Code compiles/runs
- Tests were added if code was written
- Documentation was updated
- No obvious issues introduced
"""

import json
import sys
import subprocess
from pathlib import Path
from typing import Dict, List, Optional
import re


class SubagentWorkValidator:
    def __init__(self, project_dir: str, transcript_path: str = None):
        self.project_dir = Path(project_dir)
        self.transcript_path = transcript_path
        self.issues = []
        self.suggestions = []
        
    def analyze_transcript(self) -> Dict:
        """Analyze the subagent's transcript for completeness"""
        if not self.transcript_path or not Path(self.transcript_path).exists():
            return {}
        
        try:
            with open(self.transcript_path, 'r') as f:
                content = f.read()
            
            # Look for indicators of incomplete work
            indicators = {
                'errors': len(re.findall(r'error|Error|ERROR', content)),
                'warnings': len(re.findall(r'warning|Warning|WARNING', content)),
                'todos': len(re.findall(r'TODO|FIXME|XXX', content)),
                'skipped': len(re.findall(r'skip|Skip|SKIP|defer|Defer|DEFER', content)),
                'failed_tests': len(re.findall(r'test.*fail|fail.*test|FAIL', content, re.IGNORECASE)),
            }
            
            return indicators
        except:
            return {}
    
    def check_syntax_errors(self) -> bool:
        """Check for syntax errors in recently modified files"""
        try:
            # Get recently modified files
            result = subprocess.run(
                ['find', str(self.project_dir), '-type', 'f', '-mmin', '-5'],
                capture_output=True, text=True
            )
            
            if not result.stdout:
                return True
            
            recent_files = result.stdout.strip().split('\n')
            files_with_errors = []
            
            for file in recent_files:
                if file.endswith('.py'):
                    # Check Python syntax
                    result = subprocess.run(
                        ['python', '-m', 'py_compile', file],
                        capture_output=True
                    )
                    if result.returncode != 0:
                        files_with_errors.append(Path(file).name)
                
                elif file.endswith(('.js', '.jsx')):
                    # Basic JavaScript syntax check using node
                    result = subprocess.run(
                        ['node', '--check', file],
                        capture_output=True
                    )
                    if result.returncode != 0:
                        files_with_errors.append(Path(file).name)
                
                elif file.endswith(('.ts', '.tsx')):
                    # TypeScript check if tsc is available
                    if subprocess.run(['which', 'tsc'], capture_output=True).returncode == 0:
                        result = subprocess.run(
                            ['tsc', '--noEmit', file],
                            capture_output=True
                        )
                        if result.returncode != 0:
                            files_with_errors.append(Path(file).name)
            
            if files_with_errors:
                self.issues.append(f"Syntax errors in: {', '.join(files_with_errors[:3])}")
                return False
            
            return True
        except:
            return True  # Don't block if we can't check
    
    def check_test_coverage(self) -> bool:
        """Check if tests were added for new code"""
        try:
            # Get files modified in last 5 minutes
            result = subprocess.run(
                ['find', str(self.project_dir), '-type', 'f', '-name', '*.py', '-o', '-name', '*.js', '-o', '-name', '*.ts', '-mmin', '-5'],
                capture_output=True, text=True
            )
            
            if not result.stdout:
                return True
            
            modified_code_files = []
            modified_test_files = []
            
            for file in result.stdout.strip().split('\n'):
                if file:
                    if '.test.' in file or '.spec.' in file or '/test_' in file or '/__tests__/' in file:
                        modified_test_files.append(file)
                    else:
                        modified_code_files.append(file)
            
            # If code was modified but no tests, suggest adding tests
            if modified_code_files and not modified_test_files:
                self.suggestions.append("Consider adding tests for the new code")
            
            return True  # Don't block, just suggest
        except:
            return True
    
    def check_documentation(self) -> bool:
        """Check if documentation was updated when needed"""
        try:
            # Check if any significant code files were changed
            result = subprocess.run(
                ['find', str(self.project_dir), '-type', 'f', 
                 '(', '-name', '*.py', '-o', '-name', '*.js', '-o', '-name', '*.ts', ')',
                 '-mmin', '-5'],
                capture_output=True, text=True
            )
            
            code_modified = bool(result.stdout.strip())
            
            # Check if any docs were modified
            result = subprocess.run(
                ['find', str(self.project_dir), '-type', 'f',
                 '(', '-name', '*.md', '-o', '-name', '*.mdx', '-o', '-path', '*/docs/*', ')',
                 '-mmin', '-5'],
                capture_output=True, text=True
            )
            
            docs_modified = bool(result.stdout.strip())
            
            if code_modified and not docs_modified:
                # Check if the code changes were significant (more than trivial)
                result = subprocess.run(
                    ['git', 'diff', '--stat'],
                    capture_output=True, text=True, cwd=self.project_dir
                )
                
                if result.stdout:
                    lines_changed = re.findall(r'(\d+)\s+insertion', result.stdout)
                    total_insertions = sum(int(x) for x in lines_changed) if lines_changed else 0
                    
                    if total_insertions > 20:  # Significant change
                        self.suggestions.append("Documentation may need updating for these changes")
            
            return True  # Don't block on documentation
        except:
            return True
    
    def check_imports(self) -> bool:
        """Check for missing or unused imports"""
        issues_found = False
        
        try:
            # For Python files
            result = subprocess.run(
                ['find', str(self.project_dir), '-name', '*.py', '-mmin', '-5'],
                capture_output=True, text=True
            )
            
            if result.stdout:
                for file in result.stdout.strip().split('\n'):
                    if file and Path(file).exists():
                        with open(file, 'r') as f:
                            content = f.read()
                        
                        # Check for common import issues
                        if 'import ' in content:
                            # Check for star imports (generally discouraged)
                            if re.search(r'from\s+\S+\s+import\s+\*', content):
                                self.suggestions.append(f"Avoid star imports in {Path(file).name}")
                            
                            # Check for potentially missing imports
                            used_modules = re.findall(r'\b([a-z_]+)\.\w+\(', content)
                            imported_modules = re.findall(r'import\s+(\w+)', content)
                            imported_from = re.findall(r'from\s+(\w+)', content)
                            
                            all_imports = set(imported_modules + imported_from)
                            potentially_missing = set(used_modules) - all_imports - {'self', 'cls', 'super'}
                            
                            if potentially_missing:
                                self.suggestions.append(f"Check imports in {Path(file).name}")
            
            return True  # Don't block on import issues
        except:
            return True
    
    def validate(self) -> Dict:
        """Run all validations"""
        # Analyze transcript if available
        transcript_indicators = self.analyze_transcript()
        
        # Check for serious issues
        self.check_syntax_errors()
        
        # Check for quality issues
        self.check_test_coverage()
        self.check_documentation()
        self.check_imports()
        
        # Analyze transcript indicators
        if transcript_indicators:
            if transcript_indicators.get('errors', 0) > 5:
                self.issues.append("Multiple errors detected in subagent work")
            if transcript_indicators.get('failed_tests', 0) > 0:
                self.issues.append("Tests appear to be failing")
            if transcript_indicators.get('todos', 0) > 3:
                self.suggestions.append("Several TODOs were left in the code")
        
        # Determine if we should block
        if self.issues:
            reason = "Subagent work validation failed:\n"
            reason += "\n".join(f"❌ {issue}" for issue in self.issues)
            
            if self.suggestions:
                reason += "\n\nSuggestions:\n"
                reason += "\n".join(f"💡 {suggestion}" for suggestion in self.suggestions)
            
            return {
                "decision": "block",
                "reason": reason + "\n\nPlease address these issues."
            }
        elif self.suggestions:
            # Just provide suggestions, don't block
            feedback = "Subagent work complete. Suggestions:\n"
            feedback += "\n".join(f"💡 {suggestion}" for suggestion in self.suggestions)
            
            print(feedback)
            return {}
        else:
            print("✅ Subagent work validated successfully!")
            return {}


def main():
    try:
        input_data = json.load(sys.stdin)
    except json.JSONDecodeError:
        sys.exit(1)
    
    # Handle both SubagentStop and PostToolUse(Task)
    hook_event = input_data.get('hook_event_name', '')
    
    if hook_event == 'PostToolUse':
        tool_name = input_data.get('tool_name', '')
        if tool_name != 'Task':
            sys.exit(0)
    
    project_dir = input_data.get('cwd', '.')
    transcript_path = input_data.get('transcript_path')
    
    validator = SubagentWorkValidator(project_dir, transcript_path)
    result = validator.validate()
    
    if result:
        print(json.dumps(result))
    
    sys.exit(0)


if __name__ == "__main__":
    main()